### PR TITLE
Added PittAPI computer lab functionality. Added datasets for each com…

### PR DESCRIPTION
**Pull request by SPN11**
Added functionality with @RitwikGupta 's PittAPI and associated wrapper. Added datasets for each computer lab, added fetching and processing the data, and also added filtering functionality.

Note: The PittAPI does not yet have space on Pitt's servers, so this aspect of the project will not immediately work until it is hosted on the servers (the link will not work, but I've been hosting it locally all along and it has been working).

### CODE REVIEW

**Comment by mbilker** [comment]
I am wondering if the error checking and fetching of data from an endpoint to a separate function so the basic error checking is not duplicated with non-WPRDC APIs.

File: main.js [hunk 1]
```diff
@@ -426,13 +463,101 @@
...
+            return fetch("http://labinformation.cssd.pitt.edu/lab_status/" + dataSourceName.toLowerCase())
+            //TODO: Ensure 200 response
+            .then((response) => {
+                // Inspired by https://github.com/github/fetch#handling-http-error-statuses
+                if (response.status >= 200 && response.status < 300) {
+                    return response;
+                } else {
+                    const error = new Error(response.errorText);
+                    error.response = response;
+                    displayNotification(error, "error response");
```
Comment (vonbearshark) [inline main.js]: Pull in the latest changes, we've updated how we handle this with the WPRDC function handles errors--take a look and emulate that

File: main.js [hunk 2]
```diff
@@ -426,13 +463,101 @@
...
+    }
+
     Promise.all([
         fetchWPRDCData('Police', { limit: 250 }),
         fetchWPRDCData('311', { limit: 250 }),
         fetchWPRDCData('Arrest', { limit: 250 }),
         fetchWPRDCData('Code Violation', { limit: 250 }),
         fetchWPRDCData('Library'),
-        fetchWPRDCData('Non-Traffic Violation', { limit: 250 })
+        fetchWPRDCData('Non-Traffic Violation', { limit: 250 }),
```
Comment (vonbearshark) [inline main.js]: Likewise, this all goes in one convenient function

**Comment by vonbearshark** [comment]
Yeah the `fetch` with the 200 check and `catch` is all really similar and we could probably pull that out. On the other hand, we won't need to add another one of these functions after this--it's just the WPRDC and Pitt Data (maybe one more for CMU), so I don't know if it's worth the effort right now

**Comment by mbilker** [comment]
How many issues are using datasets that aren't from WPRDC? I just want to gauge if I should work on it.

**Comment by vonbearshark** [comment]
Just the Pitt ones on the horizon right now. Don't you have an Issue assigned already @mbilker?

**Comment by mbilker** [comment]
Yes I do. I am still unsure how I should go about it.

**Comment by vonbearshark** [comment]
How is this coming along, @SPN11 ?

**Comment by SPN11** [comment]
@vonbearshark Just made the changes that emulate how `FetchWPRDCData()` handles errors and I believe my branch is fully up-to-date now

**Comment by SPN11** [comment]
Ok, everything is operating as intended, and I'm finally retrieving data. `Localhost` URL is still there, so until space on the servers is granted, these changes can't be merged...

**Comment by vonbearshark** [comment]
That's great news! Great job. Hopefully we'll have the server up and running by the end of the week, but I'll look over the rest of the code in the meantime.

File: main.js [hunk 3]
```diff
@@ -306,14 +512,125 @@
...
+                if (response.status >= 200 && response.status < 300) {
+                    return response;
+                } else {
+                    throw new Error(`Could not retrieve the ${dataSourceName} dataset; bad response.`);
+                }
+            })
+            .then((response) => response.json())
+            .then((data) => {
+                if (!data) {
+                    console.log(data);
```
Comment (vonbearshark) [inline main.js]: You can get rid of all the logging

File: main.js [hunk 4]
```diff
@@ -198,7 +198,213 @@
         notificationArea.insertBefore(box, topNotification);
     }
 
-    // data was here
+    // WPRDC data
```
Comment (vonbearshark) [inline main.js]: Didn't we move all this to the external file?

**Comment by vonbearshark** [comment]
Does this have all the most recent changes? It seems off from what we have here on master

**Comment by SPN11** [comment]
It should have all the most recent changes. It says I am 11 commits ahead of Pitt-CSC:master

**Comment by vonbearshark** [comment]
You've pulled recently? I just want to make sure we're good

**Comment by SPN11** [comment]
What seems off exactly?

**Comment by vonbearshark** [comment]
Nevermind--that's my bad. Looks great! I'll merge this in now but leave the Issue open so we can remember to update it with the new link. Thanks, @SPN11, great work! I know it was a lot of back-and-forth.

**Review by vonbearshark — APPROVED** [review]
Looks great

**Comment by SPN11** [comment]
Awesome @vonbearshark thanks! Also thank you @RitwikGupta for the guidance as well haha. Glad to be a part of it.

**Comment by RitwikGupta** [comment]
Did we have a regression? Image

**Comment by vonbearshark** [comment]
^ That makes sense because the endpoint isn't up yet.We should have more helpful errors @SPN11

**Comment by SPN11** [comment]
@vonbearshark in the `.catch(err)` in `fetchPittData` would something like `Error: Unable to fetch Pitt data` be sufficient? It's very general but narrows it down from the previous message.

**Comment by vonbearshark** [comment]
Let's use the data source name to be even more specific. Something like `Pitt dataset ${dataSourceName} could not be retrieved` for line https://github.com/SPN11/PantherView/blob/2642db420006f5e1ce1aa5b38ad49f01e84fb056/main.js#L416.

In fact, could you also change line https://github.com/SPN11/PantherView/blob/2642db420006f5e1ce1aa5b38ad49f01e84fb056/main.js#L298 to something like `Public dataset ${dataSourceName} could not be retrieved`? Does that work?

**Comment by SPN11** [comment]
That's much more descriptive, haha. Didn't know it was that easy to interpolate in JS. How is this?

@vonbearshark https://github.com/Pitt-CSC/PantherView/compare/master...SPN11:master

**Comment by vonbearshark** [comment]
Perfect! Hold off on subbing a PR until we get the server live and fix that link. Also, feel free to work on the next Pitt data Issue if you have the time!

**Comment by mbilker** [comment]
Should an actual error object be used to get a stack trace? Any way to store the original error for debugging purposes (not that WPRDC or future data sources should ever go down or change their query format)?

**Comment by vonbearshark** [comment]
We could log it maybe, but it should be in the raw network request response anyway. If we can reproduce it or if we're debugging it, we could just look there. Ideally we'd have a Dev and Prod environment setup, but I don't anticipate this really causing us headaches.

**Comment by mbilker** [comment]
@vonbearshark Alright then. Thank you.

**Comment by SPN11** [comment]
@RitwikGupta @vonbearshark So now that the server is up, would the full link for accessing lab data be:

`http://labinformation.cssd.pitt.edu/lab_status/LAWRENCE` as an example?

**Comment by RitwikGupta** [comment]
The serve is not up at the moment, but you have the right gist. The domain would *not* be labinformation... as that is owned by CSSD. It will be `http://securitas.cs.pitt.edu/something/lab_status/LAWRENCE`

**Comment by mbilker** [comment]
If that is going to be the final domain, then ask for the certificate to use SHA256 for the hash algorithm. I get a security warning by Chrome because of SHA1 being used by its certificate.

**Pull request merged**